src/snowflake/snowpark/modin/plugin/_internal/indexing_utils.py

```diff
@@ -478,34 +478,53 @@ def get_frame_by_row_pos_slice_frame(
     # Row position column required for left and right bound comparison.
     frame = internal_frame.ensure_row_position_column()
     row_pos_col = col(frame.row_position_snowflake_quoted_identifier)
-
-    # create a count column which will be used to check step
-    frame = frame.append_column("count", pandas_lit(1) + max_(row_pos_col).over())
-    count_col = col(frame.data_column_snowflake_quoted_identifiers[-1])
+    count_col_created = False
+
+    def get_count_col() -> Column:
+        nonlocal count_col_created
+        nonlocal frame
+        if not count_col_created:
+            # create a count column which will be used to check step
+            frame = frame.append_column(
+                "count", pandas_lit(1) + max_(row_pos_col).over()
+            )
+            count_col_created = True
+        return col(frame.data_column_snowflake_quoted_identifiers[-1])
 
     ordering_columns = internal_frame.ordering_columns
     start, stop, step = key.start, key.stop, key.step
 
     def make_positive(val: int) -> Column:
         # Helper to turn negative start and stop values to positive values. Example: -1 --> num_rows - 1.
-        return val + count_col if val < 0 else pandas_lit(val)
+        return val + get_count_col() if val < 0 else pandas_lit(val)
 
     step = 1 if step is None else step
+
+    # if `limit_n` is not None, use limit in the query to narrow down the search.
+    limit_n = None
+    if step < 0:
+        # Switch start and stop; convert given slice key into a similar slice key with positive step.
+        start, stop = stop, start
+    if (stop is not None and stop >= 0) and (start is None or start >= 0):
+        limit_n = stop + 1 if step < 0 else stop
+        if start is not None:
+            limit_n = limit_n - start
```

Comment:
if step is not 1, we can even further restrict the limit right? for example, if we have (1, 10, 2) we just need to restrict to 5